notion-server/src/api/mod.rs

```diff
@@ -1,30 +1,80 @@
-use crate::ethereum::Handle;
+use crate::ethereum::{args, Handle, Result as EthResult};
+use crate::Notion;
 
-use rocket::{get, routes, State};
+use rocket::response::status;
+use rocket::{config, uri};
+use rocket::{get, post, routes, State};
 
-use snafu::Snafu;
+use rocket_contrib::json::Json;
+
+use snafu::{ResultExt, Snafu};
 
 /// Shorthand for result types returned from the API server.
 pub type Result<V, E = Error> = std::result::Result<V, E>;
 
 /// Errors arising from the API server.
 #[derive(Debug, Snafu)]
-pub enum Error {}
+pub enum Error {
+    Config { source: config::ConfigError },
+}
 
-pub fn run(handle: Handle) -> Result<()> {
-    // TODO: Get some configuration options from command line.
+pub fn run(notion: &Notion, handle: Handle) -> Result<()> {
+    let config = config::Config::build(config::Environment::Development)
+        .address(format!("{}", notion.bind.ip()))
+        .port(notion.bind.port())
+        .finalize()
+        .context(Config)?;
 
-    rocket::ignite()
-        .mount("/", routes![hello])
+    rocket::custom(config)
+        .mount(
+            "/",
+            routes![
+                simulation_state,
+                show_execution_environment,
+                create_execution_environment
+            ],
+        )
         .manage(handle)
         .launch();
 
     Ok(())
 }
 
+#[tokio::main] // TODO: Check efficiency of tokio::main. Does it create or reuse thread pools?
 #[get("/")]
-fn hello(_simulation: State<Handle>) -> String {
-    // TODO: This is where we can interact with the simulation.
+async fn simulation_state(simulation: State<Handle>) -> EthResult<Json<args::SimulationState>> {
+    let state = simulation.clone().simulation_state().await?;
```

Comment:
Why clone simulation state here?  I think args:SimulationState shouldn't contain any shared data with the internal state of the simulation, so cloning seems unnecessary.  I also think if it *does* require cloning to avoid screwing up the internal state of `Simulation`, or if we're planning around that being an issue in the future, that, to me, seems like a code smell / difficult-to-use API.

Reply:
We aren't cloning the simulation here. Rather I am cloning a _handle_ to the simulation.

Remember when I said that `Handle` had to implement `Clone`? This is why.

Consider the signature of [`pub async fn send<'_>(&'_ mut self, value: T) -> Result<(), SendError<T>>`](https://docs.rs/tokio/0.2.3/tokio/sync/mpsc/struct.Sender.html#method.send). In order to send a message on a tokio channel, you need a mutable reference to the `Sender`.

It is only possible to get a shared reference (`&`) to state managed by Rocket. To get a mutable reference, I clone the `Handle` (and therefore the `Sender`.)

Reply:
ahh I see, I didn't look closely enough at the type.  I would prefer in this case we call this variable `handle` rather than `simulation` for clarity's sake.  Thoughts?